std/algorithm/iteration.d

```diff
@@ -1790,12 +1790,12 @@ if (isInputRange!Range && !isForwardRange!Range)
 {
     alias fun = binaryFun!pred;
 
-    private Range r;
+    private Range *r;
     private ElementType!Range prev;
 
-    this(Range range, ElementType!Range _prev)
+    this(ref Range range, ElementType!Range _prev)
```

Comment:
Would it simply become:  `this(return ref Range range, ElementType!Range _prev)`

Reply:
Checked my local unit test build, the iteration.d file is built with `-dip1000`. My impression was that this flag would catch needed changes. I'll wait additional clarification before making changes.

Reply:
@WalterBright Can you clarify? Should this be `return ref` as per @thewilsonator? The PR as written compiles with `-dip1000`. I searched phobos code and looked through the DIP 1000 document, but didn't see an example using `return ref` in a constructor function. I'd like to identify the correct way to write this.

Reply:
Usually, the compiler is able to infer parameter attributes like `scope` and `return` for templates, lambdas, etc. so in many case adding the attributes manually is not necessary / has no effect. However adding attributes is needed for cases where the compiler can't infer things on its own. 
Which are those cases is highly implementation specific, and it depends on a case by case basis. As rule of thumb, try to go first without `scope`/`return` and if the compiler complains, add them as necessary.
Most importantly, have test cases involving static arrays in `@safe` unittests, as the escape analysis enabled by `-dip1000` will emit errors only in `@safe` functions.

Reply:
@ZombineDev Thanks for the explanation. Current `chunkBy` unit tests are not `@safe`, they are `@system` instead. There is a separate issue listed for this, [issue 18161](https://issues.dlang.org/show_bug.cgi?id=18161). According to that report the problem is that the implementation for the forward range case uses `RefCounted`. 

My personal preference would be to leave the issue of making `chunkBy` `@safe` as a separate PR, and have this PR focus on the bug being fixed. That is, leave the implementation as written, without introducing `scope / return`. However, if reviewers prefer, I could try to make the non-forward input range `@safe` and introduce unit tests for this purpose.

Reply:
I took a closer look at this topic and experimented with creating `@safe` unit tests. Best I can tell, DIP 1000 is saying that the technique I used is not considered `@safe` under `-dip1000`. First time I've really looked at DIP 1000, but my interpretation is that it is saying that the GC can no longer be used to manage lifetime in `@safe` code, and instead some other technique, such as reference counting, must be used. Implementing this appears be an issue, perhaps temporary, as `RefCounted` is `@system` at the moment.

It'd be fantastic if I've missed something basic and there's an easy way satisfy`@safe` under `-dip1000`. If so, great. But otherwise I recommend going forward with the PR as written. The bug being addressed is quite significant, enough that it makes `chunkBy` dangerous to use. It'd be better to fix the bug even if it's `@system` under `-dip1000`. The forward input range implementation of `chunkBy` is `@system` due to use of `RefCounted`, so presumably both pieces of the `chunkBy` implementation could be made `@safe` together.